neuralmonkey/decoders/sequence_decoder.py

```diff
@@ -0,0 +1,258 @@
+        self.data_id = data_id
+        self.max_output_len = max_output_len
+        self.dropout_keep_prob = dropout_keep_prob
+
+        # TODO check the values of the parameters (vocab, ...)
+
+        with self.use_scope():
+            self.train_mode = tf.placeholder(tf.bool, [], "train_mode")
+            self.go_symbols = tf.placeholder(tf.int32, [None], "go_symbols")
+
+            self.train_inputs = tf.placeholder(
+                tf.int32, [None, None], "train_inputs")
+            self.train_mask = tf.placeholder(
+                tf.float32, [None, None], "train_mask")
+
+    @tensor
+    def batch_size(self) -> tf.Tensor:
+        return tf.shape(self.go_symbols)[0]
+
+    @tensor
+    def decoding_w(self) -> tf.Variable:
+        with tf.name_scope("output_projection"):
+            return tf.get_variable(
+                "logit_matrix",
+                [self.output_dimension, len(self.vocabulary)],
+                initializer=tf.random_uniform_initializer(-0.5, 0.5))
+
+    @tensor
+    def decoding_b(self) -> tf.Variable:
+        with tf.name_scope("output_projection"):
+            return tf.get_variable(
+                "logit_bias", [len(self.vocabulary)],
+                initializer=tf.constant_initializer(
+                    - math.log(len(self.vocabulary))))
+
+    def get_logits(self, state: tf.Tensor) -> tf.Tensor:
+        state = dropout(state, self.dropout_keep_prob, self.train_mode)
+        return tf.matmul(state, self.decoding_w) + self.decoding_b
+
+    @tensor
+    def train_logits(self) -> tf.Tensor:
+        # THE LAST TRAIN INPUT IS NOT USED IN DECODING FUNCTION
+        # (just as a target)
+        logits, _, _ = self.decoding_loop(train_mode=True)
+        return logits
+
+    @tensor
+    def train_logprobs(self) -> tf.Tensor:
+        return tf.nn.log_softmax(self.train_logits)
+
+    @tensor
+    def train_xents(self) -> tf.Tensor:
+        train_targets = tf.transpose(self.train_inputs)
+
+        return tf.contrib.seq2seq.sequence_loss(
+            tf.transpose(self.train_logits, perm=[1, 0, 2]),
+            train_targets,
+            tf.transpose(self.train_mask),
+            average_across_batch=False)
+
+    @tensor
+    def train_loss(self) -> tf.Tensor:
+        return tf.reduce_mean(self.train_xents)
+
+    @property
+    def cost(self) -> tf.Tensor:
+        return self.train_loss
+
+    @tensor
+    def runtime_loop_result(self) -> Tuple[tf.Tensor, tf.Tensor, tf.Tensor]:
+        return self.decoding_loop(train_mode=False)
+
+    @tensor
+    def runtime_logits(self) -> tf.Tensor:
+        return tuple(self.runtime_loop_result)[0]
+
+    @tensor
+    def runtime_rnn_states(self) -> tf.Tensor:
+        return tuple(self.runtime_loop_result)[1]
+
+    @tensor
+    def runtime_mask(self) -> tf.Tensor:
+        return tuple(self.runtime_loop_result)[2]
+
+    @tensor
+    def decoded(self) -> tf.Tensor:
+        return tf.argmax(self.runtime_logits[:, :, 1:], -1) + 1
```

Comment:
říct aspon v komentari, jak vziklo to peklo - zakazujem generovat padding token na indexu nula - berem jen logits od indexu jedna dál. Pak tim pádem k výsledku argmaxu musíme jedničku přičíst. Ta prostřední mínus jednička je číslo osy, podle který se dělá argmax

Reply:
DONE